### PR TITLE
rewrite workflow sharing actions into API, add Vue, add social media sharing

### PR DESCRIPTION
**Pull request by anuprulez**
This pull request rewrites the "share workflow" mako in VueJS and enables a workflow to be shareable on social media. For now, we include Twitter, Facebook, Google+ and LinkedIn where the link of any workflow can be posted through social media share option.

ssw1

ssw2

To share a workflow on multiple social media sites it uses a vue component defined in the [library](https://github.com/nicolasbeauvais/vue-social-sharing). 

Please leave us your comments and feedback!

Thanks @bgruening

### CODE REVIEW

**Comment by anuprulez** [comment]
There are 3 tests failing for API test namely: api.test_workflows.WorkflowsApiTestCase.test_run_rename_based_on_input_collection
api.test_workflows.WorkflowsApiTestCase.test_run_rename_based_on_input_conditional 
api.test_workflows.WorkflowsApiTestCase.test_workflow_metadata_validation_0

Summary of API tests:

api_test_results

But, when I run the API test locally using `./run_tests.sh -api`, all the above tests which fail at the build pass including all other API tests.

metadata
test_run

screenshot from 2018-02-13 18-01-21

I am not sure about the reason why this could happen?

**Comment by martenson** [comment]
I re-started the tests, they should pass now. It was not related to this PR.

**Comment by anuprulez** [comment]
I added a clause to show the social media share links only when a workflow is importable i.e. when it becomes accessible via a link. Thanks @bgruening for this feedback :)

@martenson does this PR look ok to you or do you have any comments? Thank you :)

**Comment by martenson** [comment]
@anuprulez I like the functionality and I will be happy to review, but am now busy with the release-related work. Afterwards I will offer feedback.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 1]
```diff
@@ -134,7 +134,7 @@ const WorkflowItemView = Backbone.View.extend({
                 this.model.id
             }">Edit</a></li><li><a href="${Galaxy.root}workflows/run?id=${this.model.id}">Run</a></li><li><a href="${
                 Galaxy.root
-            }workflow/sharing?id=${this.model.id}">Share</a></li><li><a href="${Galaxy.root}api/workflows/${
+            }workflows/share_workflow?id=${this.model.id}">Share</a></li><li><a href="${Galaxy.root}api/workflows/${
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Should probably be `workflows/share` for consistency.

File: lib/galaxy/webapps/galaxy/controllers/workflow.py [hunk 2]
```diff
@@ -319,6 +319,13 @@ def share(self, trans, id, email="", use_panels=False):
     @web.expose
     @web.require_login("Share or export Galaxy workflows")
     def sharing(self, trans, id, **kwargs):
+        """ Redirect the sharing workflow """
+        return trans.response.send_redirect(url_for('/') + 'workflows/share_workflow?id=' + id)
+
+    @web.expose
+    @web.require_login("Share or export Galaxy workflows")
+    @web.json
+    def sharing_workflow(self, trans, id, **kwargs):
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/controllers/workflow.py]: I think, we should use an api endpoint and remove this controller endpoint if possible.

File: lib/galaxy/webapps/galaxy/controllers/workflow.py [hunk 3]
```diff
@@ -472,7 +491,8 @@ def gen_image(self, trans, id):
         except Exception:
             status = 'error'
             message = 'Galaxy is unable to create the SVG image. Please check your workflow, there might be missing tools.'
-            return trans.fill_template("/workflow/sharing.mako", use_panels=True, item=stored, status=status, message=message)
+            url = url_for('/') + 'workflows/share_workflow?id=' + id + '&status=' + status + '&message=' + message
+            return trans.response.send_redirect(url)
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/controllers/workflow.py]: If we use an api endpoint, redirecting should happen through the client.

File: client/galaxy/scripts/mvc/workflow/workflow.js [hunk 4]
```diff
@@ -150,7 +150,7 @@ const WorkflowItemView = Backbone.View.extend({
                 this.model.id
             }">Run</a></li><li><a id="copy-workflow" style="cursor: pointer;">Copy</a></li><li><a class="link-confirm-shared-${
                 this.model.id
-            }" href="${Galaxy.root}workflow/sharing?unshare_me=True&id=${this.model.id}">Remove</a></li></ul>`;
+            }" href="${Galaxy.root}api/workflows/share_workflow?unshare_me=True&id=${this.model.id}">Remove</a></li></ul>`;
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow.js]: Endpoint should probably be `api/workflows/sharing` and re-use https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/base/controller.py#L1400 if possible.

File: lib/galaxy/webapps/galaxy/buildapp.py [hunk 5]
```diff
@@ -138,6 +138,7 @@ def app_factory(global_conf, load_app_kwds={}, **kwargs):
     webapp.add_client_route('/workflows/create')
     webapp.add_client_route('/workflows/run')
     webapp.add_client_route('/workflows/import_workflow')
+    webapp.add_client_route('/workflows/share')
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/buildapp.py]: Recommend this to be `workflows/sharing` for consistency. This has just recently changed.

**Comment by anuprulez** [comment]
I am not sure whether the failing API tests are related

**Comment by martenson** [comment]
I think they are, in the next 10 days there is GCC going on so responses might be delayed.

File: client/galaxy/scripts/components/ShareWorkflow.vue [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                    <p>You have not shared this workflow with any users yet.</p>
+                    <a class="action-button" :href="shareWfUrl"><span>Share with a user</span></a>
+                </div>
+            </div>
+            <div v-if="workflowItem.importable" class="sharing-section">
+                <h5>Share workflow on social media</h5>
+                <social-sharing :url="workflowUrl"
+                                title="Link to a Galaxy workflow: "
+                                quote="Galaxy workflow"
+                                hashtags="usegalaxy"
```
Comment (jxtx) [inline client/galaxy/scripts/components/ShareWorkflow.vue]: Please don't make this "hashtag" a default as it crufts up the feed.

**Comment by bgruening** [comment]
We could use `#sciworkflows`: https://twitter.com/smllmp/status/1019981835877736448

**Comment by martenson** [comment]
@galaxybot test this

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 7]
```diff
@@ -437,6 +438,69 @@ def delete(self, trans, id, **kwd):
         # TODO: Unsure of response message to let api know that a workflow was successfully deleted
         return ("Workflow '%s' successfully deleted" % stored_workflow.name)
 
+    @expose_api
+    def sharing(self, trans, id, **kwargs):
+        """
+        GET /api/workflows/sharing
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/workflows.py:467]: Clearly this method is doing more than `Return a workflow with the given id.`, so I'd like to see those things properly documented
I'm not a fan of copy-pasting controller stuff, which make very few guarantees about being stable, into the API, which should be there permanently across releases. To me all those predicates in kwargs should ideally be separate routes, and for instance returning the workflow list when requesting `unshare_me` is suboptimal.

Comment (martenson) [reply lib/galaxy/webapps/galaxy/api/workflows.py:467]: I have just started reviewing this, more changes will come. Thanks for the ideas!

**Comment by martenson** [comment]
I am moving this to 19.01, there is still work left to do with the API refactoring that I am working on but it shouldn't hold the release.

**Comment by jmchilton** [comment]
Marking as WIP this and delaying to 19.05, if it gets into 19.01 though that would be fantastic.

**Comment by anuprulez** [comment]
API test is failing for one test which passing in my local

test_nested_workflow

sdfsdf

Update: Now this test has passed :)

**Comment by anuprulez** [comment]
@martenson I rebased this branch with galaxy dev and fixed a bug related to failing selenium test. Can you say what API refactoring you wish to do for workflows? I have some time and I can work on them so that we can ship this feature with 19.01 release.

thank you for all the help.

Update: all tests passed :)

**Comment by nsoranzo** [comment]
> Update: all tests passed :)

Yeah, timeout errors in API tests are usually random and go away when re-running.

**Comment by martenson** [comment]
We discussed this privately with @anuprulez. 

I think for the social sharing part we should 
* approach this differently and reuse what @guerler added for other Galaxy objects (the Sharing.vue component from https://github.com/galaxyproject/galaxy/pull/6048)
* but make it more modular 'micro' components (like RuleComponent https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/RuleCollectionBuilder.vue#L750)

But ultimately also refactor this 'shared sharing' API endpoints to be more atomic for all actions across all objects.

Therefore we are closing this and will start work on the next iteration.

cc @jmchilton @Nerdinacan

**Pull request closed without merging**